### PR TITLE
fn eval reference docs

### PR DESCRIPTION
**Pull request by droot**
The PR adds reference docs for `kpt fn eval`.

### CODE REVIEW

**Comment by droot** [comment]
@frankfarzan this isn't ready for review yet (not able to convert it to draft for some reason). Refining the content, will be ready soon.

**Comment by droot** [comment]
@mortent I am running into this problem where `kpt fn eval --help` is printing the help twice. Any idea what could be going on ?

**Comment by droot** [comment]
fixes #1980 /cc @mengqiy

**Comment by Shell32-Natsu** [comment]
VerifyContent is still failing

**Comment by natasha41575** [comment]
@Shell32-Natsu I believe verifycontent is failing on all PRs since https://github.com/GoogleContainerTools/kpt/pull/1983

I spoke with @etefera and he said a breaking change was merged due to a typo in the testing workflow, and that Phani is addressing it

File: site/reference/fn/eval/README.md [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+description: >
+  Execute function on resources
+---
+
+<!--mdtogo:Short
+    Execute function on resources
+-->
+
+`eval` executes function on resources in a directory and writes the output to
+the local filesystem in-place. Functions are packaged as container images or
```
Comment (Shell32-Natsu) [inline site/reference/fn/eval/README.md:14]: Not only in-place? It can support out-place with `--dry-run` or `fn sink`

File: site/reference/fn/eval/README.md [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+--as-current-user:
+  Use the `uid` and `gid` of the kpt process for container function execution.
+
+--results-dir:
+  Path to a directory to write structured results. Directory must exist.
+  Structured results emitted by the functions are aggregated and saved
+  to `results.yaml` file in the specified directory.
+  If not specified, no result files are written to the local filesystem.
+  
+--dry-run:
```
Comment (frankfarzan) [inline site/reference/fn/eval/README.md:80]: i'm assuming this is being rationalized with `render` in the out-place discussion?

File: site/reference/fn/eval/README.md [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+  Path to the local directory containing resources. Defaults to the current
+  working directory. Using '-' as the directory path will cause `eval` to
+  read resources from `stdin` and write the output to `stdout`.
+```
+
+#### Flags
+
+```shell
+--image:
+  Container image of the function to execute e.g. `gcr.io/kpt-fn/set-namespace:v0.1`
```
Comment (Shell32-Natsu) [inline site/reference/fn/eval/README.md:43]: Should note that `--image` and `--exec-path` are mutually exclusive?

File: scripts/generate_site_sidebar/sidebar_template.md [hunk 4]
```diff
@@ -9,7 +9,7 @@
         - [update](reference/pkg/update/)
     - [fn](reference/fn/)
         - [render](reference/fn/render/)
-        - [run](reference/fn/run/)
+        - [eval](reference/fn/eval/)
         - [sink](reference/fn/sink/)
```
Comment (frankfarzan) [inline scripts/generate_site_sidebar/sidebar_template.md:13]: as a follow up, sink and source also be updated (they're stale and have broken links)

File: site/reference/fn/eval/README.md [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+--exec-path:
+  Path to the local executable binary to execute as a function.
+  
+--fn-config:
+  Path to the file containing `functionConfig` for the function.
```
Comment (Shell32-Natsu) [inline site/reference/fn/eval/README.md:49]: Should we clarify here that this config file will not be ignored if it's in the directory?

File: site/reference/fn/eval/README.md [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+
+--include-meta-resources:
+  If enabled, meta resources (i.e. `Kptfile` and `functionConfig`) are included
```
Comment (Shell32-Natsu) [inline site/reference/fn/eval/README.md:52]: We probably need a documentation page to list what are *meta resources*.

File: site/reference/fn/eval/README.md [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+  in the input to the function. By default it is disabled.
+
+--network:
+  If enabled, container functions are allowed to access network.
+  By default is it disabled.
```
Comment (Shell32-Natsu) [inline site/reference/fn/eval/README.md:57]: s/is it/it is/

File: site/reference/fn/eval/README.md [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+  readonly by default. Specify `rw=true` to mount volumes in read-write mode.
+
+--env:
+  List of local environment variables to be exported to the container function.
+  By default, none of local environment variables are made available to the
+  container running the function. The value can be in `key=value` format or only
+  the key of an already exported environment variable.
+
+--as-current-user:
+  Use the `uid` and `gid` of the kpt process for container function execution.
```
Comment (Shell32-Natsu) [inline site/reference/fn/eval/README.md:72]: We should document the where to use this flag. Otherwise user will find the mounted directory is not readable from the container on Linux.

File: site/reference/fn/eval/README.md (hunk 1 above)
Comment (droot) [reply site/reference/fn/eval/README.md]: Removed the `in-place` bit.

File: site/reference/fn/eval/README.md [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+
+```shell
+DIR:
+  Path to a package directory or `STDIN`.  Defaults to the current working directory.
+```
+
+#### Args
+
+```shell
+--image:
```
[the remark this answers is not in the document]
Comment (droot) [reply site/reference/fn/eval/README.md]: Done.

File: site/reference/fn/eval/README.md (hunk 3 above)
Comment (droot) [reply site/reference/fn/eval/README.md]: Added a small note in `--image` and `--exec-path` description.

File: site/reference/fn/eval/README.md (hunk 5 above)
Comment (droot) [reply site/reference/fn/eval/README.md:62]: I think we will have to figure out a better way to tell `how to ignore KRM files` for different workflows such as `live`, `packaging` and `hydration`. So not adding anything for now.

File: site/reference/fn/eval/README.md (hunk 6 above)
Comment (droot) [reply site/reference/fn/eval/README.md:69]: It is covered in `concepts` in the kpt book.

File: site/reference/fn/eval/README.md (hunk 7 above)
Comment (droot) [reply site/reference/fn/eval/README.md]: Done.

File: site/reference/fn/eval/README.md (hunk 2 above)
Comment (droot) [reply site/reference/fn/eval/README.md]: Yes.

File: site/reference/fn/eval/README.md [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+  If enabled, container functions are allowed to access network.
+  By default is it disabled.
+
+--mount:
+  List of storage options to enable reading from the local filesytem. By default,
+  container functions can not access the local filesystem. It accepts the same options
+  as specified on the [Docker Volumes] for `docker run`. All volumes are mounted
+  readonly by default. Specify `rw=true` to mount volumes in read-write mode.
+
+--env:
```
[the remark this answers is not in the document]
Comment (droot) [reply site/reference/fn/eval/README.md]: Added `-e` for this one. I checked, we don't have shorthand flags for any other flags. Will check for other `fn` commands as a follow up.

File: site/reference/fn/eval/README.md (hunk 8 above)
Comment (droot) [reply site/reference/fn/eval/README.md]: Added a note. Also added an example.

File: site/reference/fn/eval/README.md [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+```
+<!--mdtogo-->
+
+## Examples
+<!--mdtogo:Examples-->
+
+```shell
+# execute container my-fn on the resources in DIR directory and
+# write output back to DIR
+$ kpt fn --image gcr.io/example.com/my-fn eval DIR
```
[the remark this answers is not in the document]
Comment (droot) [reply site/reference/fn/eval/README.md:97]: Added it.

**Review by droot — COMMENTED** [review]
Thanks for the feedback, addressed the comments.

File: site/reference/fn/eval/README.md [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+description: >
+  Execute function on resources
+---
+
+<!--mdtogo:Short
+    Execute function on resources
+-->
+
+`eval` executes a function on resources in a directory. Functions are packaged
+as container images or binary executables.
```
Comment (frankfarzan) [inline site/reference/fn/eval/README.md:14]: I wouldn't highlight `binary executables` as that's not the main use case. It's mostly for function developers.

**Review by Shell32-Natsu — APPROVED** [review]

File: site/reference/fn/eval/README.md [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+abort and the local filesystem is left intact.
+
+Refer to the [Imperative Function Execution] for detailed overview.
+
+### Synopsis
+
+<!--mdtogo:Long-->
+
+```shell
+kpt fn eval [DIR|-] [flags] [-- fn-args]
```
Comment (frankfarzan) [inline site/reference/fn/eval/README.md:26]: do you explain fn-args anywhere here?

File: site/reference/fn/eval/README.md [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+```
+
+#### Args
+
+```shell
+DIR|-:
+  Path to the local directory containing resources. Defaults to the current
+  working directory. Using '-' as the directory path will cause `eval` to
+  read resources from `stdin` and write the output to `stdout`.
```
Comment (frankfarzan) [inline site/reference/fn/eval/README.md:35]: When using Unix pipe:
https://kpt.dev/book/04-using-functions/02-imperative-function-execution?id=chaining-functions-using-the-unix-pipe

The human-readable presentation should be sent to to `stderr` right? Currently, this doesn't display anything.

File: site/reference/fn/eval/README.md [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+```
+
+#### Flags
+
+```shell
+--as-current-user:
+  Use the `uid` and `gid` of the kpt process for container function execution.
+  Container functions are executed as `nobody` user which has very limited
```
Comment (frankfarzan) [inline site/reference/fn/eval/README.md:43]: Remove "which has very limited privileges"

File: site/reference/fn/eval/README.md [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+  privileges. You may want to use this flag to run higher privilege operations
+  such as mounting the local filesystem.
+
+--dry-run:
```
Comment (frankfarzan) [inline site/reference/fn/eval/README.md:47]: Why is this needed?

You can just do:

`eval - | <some program>`

File: site/reference/fn/eval/README.md [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+  are written to stdout. By defaults it is disabled.
+  
+--env, e:
+  List of local environment variables to be exported to the container function.
+  By default, none of local environment variables are made available to the
+  container running the function. The value can be in `key=value` format or only
+  the key of an already exported environment variable.
+
+--exec-path:
+  Path to the local executable binary to execute as a function. `eval` executes
```
Comment (frankfarzan) [inline site/reference/fn/eval/README.md:58]: Expand this to say it's for local testing purposes.

**Review by frankfarzan — CHANGES_REQUESTED** [review]

File: site/reference/fn/eval/README.md (hunk 14 above)
Comment (droot) [reply site/reference/fn/eval/README.md:35]: we are completely suppressing the human-readable presentation in this case which is not good. `stderr` could be an option, but should that be default ?

File: site/reference/fn/eval/README.md (hunk 15 above)
Comment (droot) [reply site/reference/fn/eval/README.md]: Done.

File: site/reference/fn/eval/README.md (hunk 16 above)
Comment (droot) [reply site/reference/fn/eval/README.md:52]: So I see only a narrow use-case for this that is to `execute this function, but don't have any side effects of writing to local filesystem, just show me the output`. It's like `kubectl apply --dry-run` enables you to run apply without apply the resource to cluster.

So didn't take it out. We can rationalize this as part of `out-of-place` support.

File: site/reference/fn/eval/README.md (hunk 13 above)
Comment (droot) [reply site/reference/fn/eval/README.md:26]: Added a brief description. (I need to add it's interplay with fn-config, but need to confirm that before writing it down).

File: site/reference/fn/eval/README.md (hunk 17 above)
Comment (droot) [reply site/reference/fn/eval/README.md:63]: Done.

**Review by droot — COMMENTED** [review]
Thanks for the feedback. Addressed in the last two commits.

File: site/reference/fn/eval/README.md (hunk 14 above)
Comment (frankfarzan) [reply site/reference/fn/eval/README.md:35]: what's the alternative? How unix pipes work is that by default, stdout is piped and stderr is used for observability.

File: site/reference/fn/eval/README.md (hunk 16 above)
Comment (frankfarzan) [reply site/reference/fn/eval/README.md:52]: I'd say remove this for now, unless you want to have render have dry-run as well.

**Review by frankfarzan — APPROVED** [review]
Let's follow up with open questions.

File: site/reference/fn/eval/README.md (hunk 14 above)
Comment (droot) [reply site/reference/fn/eval/README.md:35]: ok, then we can use stderr.
